File: solutions/3147. Taking Maximum Energy From the Mystic Dungeon/3147.py
# Time complexity: O(n)
# Space complexity: O(n)
class Solution:
  def maximumEnergy(self, energy: List[int], k: int) -> int:
    dp = energy[:]
    for i in reversed(range(len(energy) - k)):
      dp[i] += dp[i + k]

    return max(dp)

File: solutions/3147. Taking Maximum Energy From the Mystic Dungeon/3147.cpp
// Time complexity: O(n)
// Space complexity: O(n)
class Solution {
 public:
  int maximumEnergy(const vector<int>& energy, const int k) {
    vector<int> dp(energy);
    for (int i = energy.size() - 1 - k; i >= 0; --i)
      dp[i] += dp[i + k];

    return ranges::max(dp);
  }
};